nbextensions/appCell2/widgets/appCellWidget.js

```diff
@@ -1317,24 +1319,46 @@ define(
             }
 
             function doJobStatus(message) {
-                const existingState = model.getItem('exec.jobState'),
-                    existingJobId = existingState
-                        ? existingState.job_id
+                const existingJobState = model.getItem('exec.jobState'),
+                    existingJobId = existingJobState
+                        ? existingJobState.job_id
                         : model.getItem('exec.launchState.job_id'),
                     newJobState = message[existingJobId].jobState,
-                    { outputWidgetInfo } = message[existingJobId],
-                    forceRender =
-                        (!existingState || !Jobs.isValidJobStateObject(existingState)) &&
-                        Jobs.isValidJobStateObject(newJobState);
-                if (!existingState || !_.isEqual(existingState, newJobState)) {
+                    { outputWidgetInfo } = message[existingJobId];
+                if (!existingJobState || !_.isEqual(existingJobState, newJobState)) {
                     model.setItem('exec.jobState', newJobState);
                     if (outputWidgetInfo) {
                         model.setItem('exec.outputWidgetInfo', outputWidgetInfo);
                     }
+                    // update the job state
+                    updateJobState();
                 }
                 model.setItem('exec.jobStateUpdated', new Date().getTime());
 
-                updateFromJobState(newJobState, forceRender);
+                const newFsmState = (function () {
+                    switch (newJobState.status) {
+                        case 'created':
+                        case 'estimating':
+                        case 'queued':
+                            return fsmState.PROCESSING_QUEUED;
+                        case 'running':
+                            return fsmState.PROCESSING_RUNNING;
+                        case 'completed':
+                            stopListeningForJobMessages();
+                            return fsmState.COMPLETED;
+                        case 'terminated':
+                            stopListeningForJobMessages();
+                            return fsmState.TERMINATED;
+                        case 'error':
+                        case 'does_not_exist':
+                            stopListeningForJobMessages();
+                            return fsmState.RUNTIME_ERROR;
+                        default:
+                            throw new Error('Invalid job state ' + newJobState.status);
+                    }
+                })();
+                fsm.newState(newFsmState);
+                updateState();
```

Comment:
there was a line here that ran `initializeFsm()` if `forceRender` was true, and I cannot work out why you would want to do that -- as it was, `initializeFsm()` was setting the wrong FSM state because it uses the value stored in `model.getItem('fsm.currentState')`, which lags behind what's in the FSM's `currentState`.